docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+The components are installed in a proprietary fashion by applying highly customized Kubernetes YAML resources. They are hard to customize as the spec is proprietary and hard to extend because each customization must be passed through configuration to the end-user. Changes to this spec must be documented and are less discoverable. This method is not supported by the upstream which makes it difficult to pin addon installation issues on the upstream project and difficult to request support from the upstream.
+
+
+## Decision
+
+We will develop a Kubernetes [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) in _Go_ that will install, configure and manage add-on applications in the cluster. The operator will extend the Kubernetes API through Custom Resource Definitions (CRD) and will allow users to declaritively manage add-ons with a Custom Resource (CR). 
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
```

Comment:
We cannot make a controller create a another controller. I think what you want to say would be like
```suggestion
The solution involves implementing a minimum of one controller, the Installer controller, to periodically reconcile the Installer custom resource (CR) with the cluster. The Installer controller will delegate the management of add-ons and other required processes, such as Storage Migrations, to separate controllers. During reconciliation, the main controller will determine the add-ons and processes specified in the Installer CR and create individual controllers for each of them. This approach ensures that each component, be it an add-on or a required process, is managed by a dedicated controller, following the concept of single responsibility.
```

OR 

```suggestion
The solution involves implementing a minimum of one controller, the Installer controller, to periodically reconcile the Installer custom resource (CR) with the cluster. The Installer controller will not only handle the reconciliation process but also create custom resources (CRs) to represent the add-ons and processes, such as Storage Migrations, within the cluster.

During reconciliation, the Installer controller will determine the add-ons and processes specified in the Installer CR. It will then create separate CRs to represent each add-on or process. For example, if there is a need to perform a Storage Migration, the Installer controller will create a specific CR to represent that process.

In addition to creating the CRs, the Installer controller will delegate the management of each CR to dedicated controllers. These controllers will be responsible for monitoring and maintaining the desired state of their associated CRs, following the concept of single responsibility.

By adopting this approach, the operator can ensure that the Installer controller creates and manages the necessary CRs, representing add-ons and processes like Storage Migrations, while allowing separate controllers to handle the management and enforcement of the desired state for each specific component.
```